---
output: 
    html_document:
        code_folding: hide
        df_print: tibble
        highlight: default
        theme: paper
        toc: true
        toc_depth: 5
        toc_float: true
        css: src/assets/style/style.css

always_allow_html: yes


---

```{r setup, include=FALSE}
knitr::opts_chunk$set(
    tidy = F,
    tidy.opts = list(width.cutoff = 120),
    message = FALSE,
    warning = FALSE
)

# module purge;source /camp/stp/babs/working/software/modulepath_new_software_tree_2018-08-13;module load pandoc/2.2.3.2-foss-2016b;ml R/4.0.3-foss-2020a

if (!require("remotes")){
  install.packages("remotes")
}

remotes::install_github("rstudio/renv")

if (!file.exists("renv.lock")){
    renv::init()
} else {
    renv::restore(prompt=FALSE)
}

workdir <- getwd()
figureCount <- 1
tableCount <- 1
upload.to.database <- TRUE

```
   
    
```{r, echo=T, eval=TRUE, warning=FALSE, results="asis"}

###############################################################################
## Part 1: Set essential parameters (essential section)                      ##

## Essential parameters
pipelineList <- list()

## Set folder
pipelineList[["folder"]] <- "/camp/stp/babs/working/boeings/Projects/anastasioud/frederick.clasen/477_DAL_bulkRNASeq_RN21224/"

## Path were the biologic spec file will be kept:
pipelineList[["biologicSettingsFN"]] <- paste0(pipelineList[["folder"]], "workdir/bulkRNAseq_workflow/design/biologic.settings.file.csv")

## Project parameters
pipelineList[["project_id"]] <- "RN21224"
pipelineList[["lims.id"]] <- "RN21224"

pipelineList[["labname"]] <- "Anastasiou"

pipelineList[["NtopGenes"]] <- 500

# options: "bulk_rna_seq", "sc_rna_seq"
pipelineList[["experiment.type"]] <- "bulk_rna_seq"  
pipelineList[["species"]] <- "mus_musculus"
pipelineList[["release"]] <- "release-89"


pipelineList[["timecourse.units"]] <- ""
pipelineList[["count.table.headline"]] <- "TPM Values for all Samples"
pipelineList[["count.table.sidelabel"]] <- "TPM"
pipelineList[["heamap.headline.text"]] <- "Heatmap: Row-averaged Expr"


## Reference data selections
pipelineList[["Database Parameters	"]] <- ""

## Enter lab database name here
pipelineList[["primDataDB"]] <- "dal_data"

## These three can be left pre-set
pipelineList[["ref.cat.db"]] <- "reference_categories_db_new"
pipelineList[["db.user"]] <- "babs"
pipelineList[["host"]] <- "10.27.241.82"

## Add lab categories table, if available.
pipelineList[["lab.categories.table"]] <- "da_lab_categories"

## Set additional reference tables ##

dfRef <- cbind(
    c("Hallmark Signatures",
      "Pathways",
    "GO-BP",
    "GO-MF",
    "TF Motifs",
    "Protein Complexes",
    "Immunologic Signatures",
    "LINCS Down",
    "LINCS Up",
    "Cell Type Signatures",
    "Cell Type Signatures"
    ),
    c("Signatures	mysigdb_h_hallmarks",
      "mysigdb_c2_1329_canonical_pathways",
      "mysigdb_c5_BP",
      "mysigdb_c5_MF",
      "TRANSFAC_and_JASPAR_PWMs",
      "networkcategories",
      "mysigdb_c7_immunologic_signatures",
      "LINCS_L1000_Chem_Pert_down",
      "LINCS_L1000_Chem_Pert_up",
      "mysigdb_sc_sig",
      "cibersort_L22"
    )
)

dfRef <- cbind(
    rep("referenceTableListDB", nrow(dfRef)),
    dfRef
)



pipelineList[["ReferenceTable"]] <- dfRef

## Parameters that might be set automatically at the Crick, when information
# is available via asf. 

## This is the project under which this experiment will be organised in the db
pipelineList[["project_name"]] <- "Dietary Effects on Liver Metabolism"

pipelineList[["Documentation Parameter"]] <- ""
pipelineList[["title"]] <- "Automatic"
pipelineList[["subtitle"]] <- "Automatic"
pipelineList[["abstract"]] <- "Automatic"

## End Part 1                                                                ##
###############################################################################

###############################################################################
## Option: create design file for Crick experiment                           ##
## Requirement: sample.ids.txt file with columns sampleID, sample.id

if (dir.exists("/Volumes/babs/working/boeings/")){
    hpc.mount <- "/Volumes/babs/working/boeings/"
} else if (dir.exists("Y:/working/boeings/")){
    hpc.mount <- "Y:/working/boeings/"
} else if (dir.exists("/camp/stp/babs/working/boeings/")){
    hpc.mount <- "/camp/stp/babs/working/boeings/"
} else {
    hpc.mount <- ""
}

baseMount <- gsub(
    "boeings/",
    "",
    hpc.mount
)

pipelineList[["pathToSeqStorageFolder"]] <- "/camp/stp/sequencing/outputs/babs/data/anastasioud/jack.carruthers/RN21224/primary_data/220218_A01366_0146_BH757VDMXY/fastq/"

pipelineList[["fastqDir"]] <- paste0(
    pipelineList[["folder"]],
    "FASTQ_files/"
)

pipelineList[["localWorkDir"]] <- paste0(
    pipelineList[["folder"]],
    "workdir/"
)

returnList <- biologicSeqTools2::organizeFastqFiles(
    baseMount = "",
    pathToSeqStorageFolder = pipelineList[["pathToSeqStorageFolder"]],
    fastqOutputDir = pipelineList[["fastqDir"]],
    localWorkDir = pipelineList[["localWorkDir"]],
    checkFileExists = FALSE
)

shellscriptVec <- returnList$shellscriptVec
concatenationRequired <- returnList$concatenationRequired

if (concatenationRequired){
    print("Concatenation script executed?")
    pathToSeqStorageFolder <- pipelineList[["fastqDir"]]
} else {
    pathToSeqStorageFolder <- pipelineList[["pathToSeqStorageFolder"]]
}

#################################################################################
## If concatenation required, generate new input files                         ##
setwd(pipelineList[["localWorkDir"]])
if (concatenationRequired){
    system("concatFASTQfiles.sh ")
}

## Temporary fix
setwd(paste0(pipelineList[["localWorkDir"]], "bulkRNAseq_workflow"))
## end

## Done                                                                        ##
#################################################################################

# ###############################################################################
# ## Create Design File                                                        ##
#
## this setup requires a file named sample.ids.txt in workir with two columns:
# sampleID (with the ASF id) and 
# sample.id with the experimental sample name. 

pipelineList[["paired.end"]] <- FALSE
pipelineList[["baseDesignFN"]] <- paste0(
  pipelineList[["localWorkDir"]],
  "sample.ids.txt"
) 

dfDesign <- biologicSeqTools2::createDesignFileCrickASFsamples(
    pathToSeqStorageFolder = pathToSeqStorageFolder,
    FNsampleAnnotation = pipelineList[["baseDesignFN"]],
    paired.end = pipelineList[["paired.end"]] ,
    baseMount = "",
    fastqDir = pipelineList[["fastqDir"]]

)


##
###############################################################################

###############################################################################
## Create a design file                                                      ##

# The easiest is create a basedesign file with the following columns:
# |sample.id|sample.group|dataseries|comp_1|comp_2|...|comp_N|LRT_Treatment|LRT_...|
# |f_experimental_factor_1|f_experimental_factor_2|...

# Load design file from file:

# baseDesignFN <- paste0(
#   pipelineList[["folder"]],
#   "workdir/bulkRNAseq_workflow/design/design.file.txt"
# )
#   
#  
# 
# dfDesign <- read.delim(
#     baseDesignFN, 
#     header = T,
#     sep = "\t",
#     stringsAsFactors = F
# )


## Check: Make sure the above required columns are present:
names(dfDesign)

## add additional columns
# The function below requires a sample name structure of 
# dataseries_sampleGroup_replicate. 
# There should be no underscores, but the two in the above example. 
# dfDesign <- biologicSeqTools2::completeDesignBasedOnSampleID(dfBaseDesign)

## Make sure colors are specified
dfDesign$sample.groups <- NULL

###############################################################################
## Add sample group colors
pos <- grep("sample.group_color", names(dfDesign))
        
if (length(pos) == 0){
    sample.group <- unique(dfDesign$sample.group)
    sample.group_color <- sample.group


            #library(scales)
    sample.group_color = scales::hue_pal()(length(sample.group_color))
            #sample.group_color = c("#990000", "#009900")

    dfGroupColors <- unique(data.frame(sample.group, sample.group_color))

    dfDesign <- merge(dfDesign, dfGroupColors, by.x = "sample.group", "sample.group")

}
##
###############################################################################

pos <- grep("^timepoint$", names(dfDesign))

if (length(pos) == 1){
    library(dplyr)
    dfDesign <- dfDesign %>%
        mutate(timepoint = as.numeric(timepoint)) %>%
        arrange(dataseries, timepoint, f_replicate, .by_group = TRUE)  

    dfDesign <- data.frame(dfDesign)
}
## Done with colors
##
###############################################################################



# Location where the design file will be saved
pipelineList[["designFN"]] <- paste0(pipelineList[["folder"]], "workdir/bulkRNAseq_workflow/design/", pipelineList[["project_id"]],".design.table.txt")

## ensure a replicate factor column is present
# dfDesign[["f_replicate"]] <- paste0("R", sapply(dfDesign$sample.id, function(x) unlist(strsplit(x, "_"))[3]))


write.table(
    dfDesign,
    pipelineList[["designFN"]] , 
    sep = "\t",
    row.names = F
)

## check design file ##
## This section will check if all relevant columns are present
names(dfDesign)

# sample.id has a unique name for each sample without spaces, 
# e.g. dataseries_sampleGroup_rep1

## sample.group summarises sample.ids of the same sample group, 
# e.g. dataseries_sampleGroup

## dataseries
# supergroup. E.g. Treatment over various timepoints or a cell line

## comp_1 to comp_N-
# comp_1 to comp_N specify differential gene expression comparisons. In each column
# samples are assigned to group_A or group_B prefixed with a 1_ or 2_ depending on the 
# position in the A vs B differential gene expression comparison.

## LRT_....
# Saple groups for a LRT comparison. Assign each sample to an LRT group or leave 
# empty if that sample is not to be included into that LRT. 

## timeseries
# if you have a time-series experiment, indicate a timepoint for each sample. 

## f_
# indicate experimental factors in the experiment. Prefix each variable factor
# with an f_

## Keep in mind that all entries in the DESeq2 models need to be present as columns


## Done design file                                                          ##
###############################################################################

###############################################################################
## Create model files                                                        ##

# If you already have a model file, just add the path to the file at the bottom
# of this section. There is no need to create the file. 
# comparisonID <- names(dfDesign)[grep("comp_", names(dfDesign))]
# comparison <- rep("", length(comparisonID))
# for (i in 1:length(comparisonID)){
#     contrasts = sort(as.vector(unique(dfDesign[,comparisonID[i]])), decreasing = FALSE)
#     contrasts = contrasts[contrasts != ""]
#     
#     contrasts <- gsub("^1_", "", contrasts)
#     contrasts <- gsub("^2_", "", contrasts)
#     
#     #Create contrast vector
#     #contrast.vector = c([condition],[1_diff.gene set, e.g. mt],[2_baseline, e.g. wt])
#     #if (contrasts[2] != "scr"){
#     #  contrasts = rev(contrasts)
#     #}
#     sel.col = contrasts
#     
#     contrast.vector = append("condition", contrasts)
#     colName = paste(contrasts, collapse = "_vs_")
#     comparison[i] <- colName
# }
# 
# test <- rep("Wald", length(comparison))
# type <- rep("DGE", length(comparison))
# 
# model <- rep("~ condition", length(comparison))
# 
# ## You might need to replace the above model vector with a bespoke model 
# # for each comparison
# # example for an experiment with 5 comparisons
# # model <- c(
# #     "~ condition",
# #     "~ condition + f_timepoint + condition:f_timepoint",
# #     "~ condition + f_timepoint + condition:f_timepoint",
# #     "~ condition + f_timepoint + condition:f_timepoint",
# #     "~ condition + f_timepoint + condition:f_timepoint"
# # )
# 
# ## Reduced model not relevant for a DGE, but a column that needs to be present
# reducedModel <- rep("", length(comparison))
# 
# ## Normalise all samples together, or just those involved in the DGE comparison
# normalizeAllSamplesTogether <- rep(FALSE, length(comparison))
# betaPrior <- rep(TRUE, length(comparison))
# 
# 
# dfModel <- data.frame(
#     comparison,
#     comparisonID,
#     test,
#     type,
#     model,
#     reducedModel,
#     normalizeAllSamplesTogether,
#     betaPrior, stringsAsFactors = TRUE
# )
# 
# 
# ## Now add the LRT part 
# comparisonID <- names(dfDesign)[grep("LRT_", names(dfDesign))]
# comparison <- names(dfDesign)[grep("LRT_", names(dfDesign))]
# 
# 
# if (length(comparison) > 0){
#         test <- rep("LRT", length(comparison))
#         type <- rep("LRT", length(comparison))
#         model <- c(
#             "~ LRT_Sample_Treatment",
#             "~ LRT_timepoint"
#         )
#         # 
#         # 
#         reducedModel <- c(
#             "~ 1",
#             "~ 1"
#         )
#         # 
#         normalizeAllSamplesTogether <- rep(FALSE, length(comparison))
#         betaPrior <- rep("", length(comparison))
#         
#         
#             
#         dfMLRT <- data.frame(
#             comparison,
#             comparisonID,
#             test,
#             type,
#             model,
#             reducedModel,
#             normalizeAllSamplesTogether,
#             betaPrior,
#             stringsAsFactors = FALSE
#         )
#         
#         dfModel <- rbind(
#             dfModel,
#             dfMLRT
#         )
# }
# 
# dfModel[is.na(dfModel)] <- ""

## Here model is loaded from file
 
pipelineList[["modelFN"]] <- paste0(pipelineList[["folder"]], "/workdir/bulkRNAseq_workflow/design/", pipelineList[["project_id"]],".model.table.txt")
#model.file <- "design/dge.model.table.txt"

file.exists(pipelineList[["modelFN"]])

## Write model to file
#write.table(dfModel ,pipelineList[["modelFN"]] , row.names = FALSE, sep = "\t")

## Done model file                                                           ##
###############################################################################

###############################################################################
## Part 2A: Use to start from fastq files                                    ##

###############################################################################
## Create analysis scripts                                                   ##
# sDir <- getwd()
# setwd(Obio@parameterList$localWorkDir)
# 
# Obio <- biologicSeqTools2::createbulkRNASeqAnalysisBashScripts(
#     obj = Obio,
#     scriptVecSlot = "scriptVec"
# )
# setwd(sDir)

## Run analysis BASH script ##
## Done creating analysis bash scripts                                       ##
###############################################################################

## End Part 2A                                                               ##
###############################################################################

###############################################################################
## Part 2B: Start from (RSEM) read count matrix                              ##

## Create RSEM count matrix if it doesn't exist yet

## This is the folder in which the [sample.id].genes.result files live
RSEM_folder <- paste0(pipelineList[["folder"]], "workdir/RSEM/Ensembl")
pipelineList[["RSEM_folder"]]  <- RSEM_folder 

## This is the name of the RSEM output file
countTableFN <- paste0(pipelineList[["folder"]], "workdir/RSEM/RN21224.count.data.txt")


## Save as parameter
pipelineList[["countTableFN"]] <- countTableFN


files <- paste0(
    RSEM_folder, "/",
    list.files(RSEM_folder )[grep(".genes.results", list.files(RSEM_folder))]
)

files = paste(
    files,
    collapse = " "
)

RSEM.CMD = paste(
    "module load RSEM/1.3.0-foss-2016b;",
    "rsem-generate-data-matrix ",
    files,
    " > ",
    pipelineList[["countTableFN"]],
    sep=""
)

## Run RSEM command
## The above command might only work on the command line
# system(RSEM.CMD)

## Read resulting RSEM table
dfRSEM <- read.delim(
    pipelineList[["countTableFN"]], 
    sep = "\t",
    header = T,
    stringsAsFactors = F
)

## Make sure it's integers:
dfRSEM[,2:ncol(dfRSEM)] <- round(dfRSEM[,2:ncol(dfRSEM)])

names(dfRSEM) <- gsub(paste0("X", gsub("/", ".", pipelineList[["folder"]]), ".workdir.RSEM.Ensembl."), "", names(dfRSEM))
names(dfRSEM) <- gsub(".genes.results", "", names(dfRSEM))

## Format: |gene_id column names X|sample.id[1]|sample.id[2]|sample.id[3]|...

write.table(dfRSEM, pipelineList[["countTableFN"]], sep = "\t", row.names=F)


## Format of the RSEM read count matrix:

## TPM table ##
## Option A: Specify the folder in which the RSEM genes.result files reside

## Option B: Specify TPM file directly. The TPM file needs to have the following 
##           format: 

## End Part 2A                                                               ##
###############################################################################

###############################################################################
## Part 2C: I have a read-count matrix, a TPM file, and DESeq2 differential  ##
##          gene expression outputs                                          ##

## RSEM_folder
## The RSEM folder points to the folder containing the [sampleID].genes.result files
pipelineList[["RSEM_folder"]] <- paste0(pipelineList[["folder"]], "workdir/RSEM/Ensembl/")


## TPM file ##
## The tpm file should contain the following columns:
## first column: primary gene alignment ID, named either "X" or "ENSG|ENSMUSG|or whatever species
## second to N-th column: one column for each sample as liste in Obio@dfDesign$sample.id. Can be extended by a
## _TPM suffix. 
pipelineList[["TpmTableFN"]] <- paste0(
        pipelineList[["folder"]],
        "basedata/",
        pipelineList[["projectID"]],
        ".tpm.file.txt"
        
    )

if (!is.null(pipelineList[["RSEM_folder"]])){
    files <- list.files(pipelineList[["RSEM_folder"]] )
    files <- files[grep("genes.results", files)]
    samples <- gsub(".genes.results", "", files)
    
    files <- paste0(pipelineList[["RSEM_folder"]] , files)
    
    
    
    dfTables <- biologicSeqTools2::create.tpm.and.fpkm.tables(
        files = files,
        samples = samples
    )
    
    dfTPM <- dfTables$df.tpm
    
    
    write.table(
        dfTPM,
        pipelineList[["TpmTableFN"]],
        sep = "\t",
        row.names = F
    )
}




#tpmFN <- "/camp/stp/babs/working/bioinformatics/projects/sahaie/colin.ratcliffe/Capivasertib_resistant_in_triple_negative_breast_tumour_organoids/babs/outputs/data/rsem.merged.gene_tpm.tsv"

dfTPM <- read.delim(
    pipelineList[["TpmTableFN"]], 
    sep = "\t",
    stringsAsFactors = F,
    header=T
)

pipelineList[["primaryAlignmentGeneID"]] <- gsub("[0-9]", "", as.vector(dfTPM$gene_id[1]))

dfTPM$gene_name <- NULL
dfTPM$transcript_id.s. <- NULL
names(dfTPM) <- gsub("gene_id", pipelineList[["primaryAlignmentGeneID"]] , names(dfTPM))


write.table(dfTPM, pipelineList[["TpmTableFN"]] , sep="\t", row.names=F)


## End Part 2C                                                               ##
###############################################################################

###############################################################################
## Create settings file                                                      ##

###############################################################################
## Create parameter File                                                     ##

## To create pipelineJSON, remove reference list ##
dfRefTab <- data.frame(t(pipelineList[["ReferenceTable"]]))
pipelineList[["ReferenceTable"]] <- NULL

pipelineJson <- jsonlite::toJSON(pipelineList)

dfObio <- as.data.frame(jsonlite::fromJSON(pipelineJson))
dfObio <- rbind(names(dfObio), dfObio)
colnames(dfObio) <- NULL
dfObio <- rbind(dfObio, rep("", ncol(dfObio)))

dfObio <- cbind(
    dfObio, 
    dfRefTab
)
    

dfObio <- data.frame(t(dfObio))

settingsFN <- paste0(pipelineList[["folder"]], "workdir/bulkRNAseq_workflow/design/biologic.settings.csv")
write.table(dfObio, settingsFN , row.names=FALSE, col.names=FALSE, sep=",")

## Done 
###############################################################################

## Done                                                                      ##
###############################################################################

## Settings file created and saved in design/biologic.settings.csv

```    
    

<!-- Essential 1: (output required later): Set Parameters and directories and create Obio object Chapter/Module -->
<!-- This section creates the plain Obio object without any data in it. -->
```{r child = 'src/modules/settings/set.parameters.Rmd', eval=TRUE}
```

<!-- Create fastq analysis files-->
```{r, echo=T, eval=TRUE, warning=FALSE, results="asis"}

dfDesign[["FASTQ"]] <- dfDesign$sample.id
Obio@dfDesign <- dfDesign


###############################################################################
## Create analysis scripts                                                   ##

Obio@parameterList[["stranded"]] <- TRUE
Obio@parameterList[["scriptVec"]] <- as.vector(NULL, mode="character")
Obio@parameterList[["AlignFASTQcolumn"]] <- "sample.id"

Obio@parameterList[["ModuleFASTQC"]] = "module load FastQC/0.11.5-Java-1.8.0_92"
Obio@parameterList[["ModuleTrimGalore"]] = "Trim_Galore/0.4.2-foss-2016b"
Obio@parameterList[["TrimGaloreMinLength"]] = 25
Obio@parameterList[["TrimGaloreMinQuality"]] = 20
Obio@parameterList[["read.length"]] =  "100bp"

#Obio <- setMountingPoint(Obio)
#Obio <- setAnalysisPaths(Obio)

## Set genome paths
Obio <- biologicSeqTools2::setCrickGenomeAndGeneNameTable(Obio)
#Obio <- createAnalysisFolders(
#    Obio
#)
#Obio <- setDataBaseParameters(Obio)

## This can be upgraded to web retrieval of annotation data 
#Obio <- addGeneAnnotation(Obio)

sDir <- getwd()
setwd(Obio@parameterList$localWorkDir)


Obio <- biologicSeqTools2::createbulkRNASeqAnalysisBashScripts(
    obj = Obio,
    scriptVecSlot = "scriptVec"
)
setwd(sDir)

## Run analysis BASH script ##
## Done creating analysis bash scripts                                       ##
###############################################################################
```

<!-- Essential 2: Load all available data into Obio object and check -->
<!-- This part will move to part B -->
```{r child = 'src/modules/part_A/A.1.add.data.to.analysis.object.Rmd', eval=FALSE}
```

<!-- Essential 1: (output required later): Set Parameters and directories and create Obio object Chapter/Module -->
<!-- This part will move to part B -->



```{r saveobject, eval=TRUE, echo=T, results=F}
### Will save Obio object here, so it can be re-used with different parameters
save(Obio, 
     file = paste0(
         Obio@parameterList$localWorkDir,
         Obio@parameterList$project_id,
         ".bioLOGIC.Robj"
     )
)

print("R bioLOGIC bulk RNAseq Object initialized.")

# renv::snapshot()

```

## Documentation
```{r documentation, eval=TRUE, echo=T, results=T}
sessionInfo()
```

```{r create_report_params, eval=T, results="asis"}

## Try to retrieve project data from db ##
db.pwd2 <- "zU3ufd9L"
db.user2 <- "reader"
host2 <- "clvd1-db-u-p-17.thecrick.org"
projectParams <- Obio@documentationParams

tryCatch({
    dbDB = DBI::dbConnect(
        drv = RMySQL::MySQL(), 
        user = db.user2, 
        password = db.pwd2, 
        host = host2, 
        dbname = "clarity_shadow"
    )
    dfProposal <-  DBI::dbGetQuery(
        dbDB, 
        paste0("SELECT * FROM clarify_asf_proposals WHERE project_name ='",Obio@projectDetailList$lims.id,"'")
    )
    dbDisconnect(dbDB)
}, error = function(x) {
    message("Project Database could not be reached or has no entry in Obio@parameterList$lims.id for this analysis.")
})

###############################################################################
## Helper
firstup <- function(x) {
    substr(x, 1, 1) <- toupper(substr(x, 1, 1))
    x
}
##
###############################################################################


if (exists("dfProposal")){
    if (!is.na(dfProposal[1,"ProjectAlias"]) & dfProposal[1,"ProjectAlias"] != ""){
        projectParams[["title"]] = paste0(dfProposal[1,"ProjectAlias"], " - ", dfProposal[1,"project_name"])
    }
    
    if (!is.na(dfProposal[1,"project_user"]) & dfProposal[1,"project_user"] != ""){
        labString <- firstup(dfProposal[1,"user_lab"])
        labString <- substr(labString, 1, (nchar(labString) - 1))
      
        projectParams[["subtitle"]] = paste0(labString, " Lab - ", dfProposal[1,"project_user"])
        projectParams[["subtitle"]] <- gsub("^ Lab - ", "", projectParams[["subtitle"]])
        
    }
    
    if (!is.na(dfProposal[1,"proposal_text"]) & dfProposal[1,"proposal_text"] != ""){
        projectParams[["abstract"]] = dfProposal[1,"proposal_text"]
       
        
    }
}
   
## Escape all special characters
projectParams <- lapply(
  projectParams, function(x) 
  #gsub("([.|()\\^{}+$*?]|\\[|\\])", "\\\\\1", x)
  gsub("([.|()/\\^{}+$*?]|\\[|\\])", " ", x)
) 

projectParams <- lapply(
  projectParams, function(x) 
  #gsub("([.|()\\^{}+$*?]|\\[|\\])", "\\\\\1", x)
  gsub("\\\n", " ", x)
) 


#projectParams$title <- "Title"
projectParams$abstract <- "This is the QC section. In this section basic sequencing parameters are documented. Then particles that contain only ambient RNA or dead/dying cells are identified to be removed. An estimate on doubets (GEMS containing more than one cell) is made. Finally biological biases by cell cycle or gender are evaluated and regressed out as required."
#projectParams$subtitle <- "Abstract"

```

---
title: "`r projectParams$title`"
subtitle:  "`r projectParams$subtitle`"
author:
    - Stefan Boeing^[The Francis Crick Institute, stefan.boeing@crick.ac.uk]
date: 'Compiled: `r format(Sys.Date(), "%B %d, %Y")`'

abstract: |
    "`r projectParams$abstract`"

---