---
output: 
    html_document:
        code_folding: hide
        df_print: tibble
        highlight: default
        theme: paper
        toc: true
        toc_depth: 5
        toc_float: true
        css: src/assets/style/style.css

always_allow_html: yes
---

```{r setup, include=FALSE}
###############################################################################
## Recommended R-version                                                     ##

# module purge;source /camp/stp/babs/working/software/modulepath_new_software_tree_2018-08-13;module load pandoc/2.2.3.2-foss-2016b;ml R/4.0.3-foss-2020a

## Done                                                                      ##
###############################################################################


###############################################################################
## Set the environment                                                       ##

if (!require("remotes")){
  install.packages("remotes")
}

remotes::install_github("rstudio/renv")

if (!file.exists("renv.lock")){
    renv::init()
} else {
    renv::restore(prompt=FALSE)
}


renv::install("bioc::DESeq2")
renv::install("bioc::clusterProfiler")
renv::install("decusInLabore/biologicSeqTools2")

## Done                                                                      ##
###############################################################################


###############################################################################
## Set knitr options                                                         ##

knitr::opts_chunk$set(
    tidy = F,
    tidy.opts = list(width.cutoff = 120),
    message = FALSE,
    warning = FALSE
)

##                                                                           ##
###############################################################################

###############################################################################
## Set global variables                                                      ##

cDir <- getwd()
setwd("..")
workdir <- getwd()
setwd(cDir)

figureCount <- 1
tableCount <- 1
upload.results.to.database <- TRUE
shinyBaseServerURL <- "shiny-bioinformatics.crick.ac.uk"

## Done                                                                      ##
###############################################################################


###############################################################################
## Load biologic object from part A                                          ##
library(biologicSeqTools2)
FN <- paste0("../", list.files("..")[grep("bioLOGIC.Robj", list.files(".."))])
load(FN)
## Done                                                                      ##
###############################################################################

```


```{r , echo=TRUE, eval=T, warning=FALSE}


###############################################################################
##                                                                           ##

# FN <- paste0(hpc.mount, "Projects/reference_data/documentation/BC.parameters.txt")
# dbTable <- read.delim(
#     FN, 
#     sep = "\t",
#     stringsAsFactors = F
# )
# db.pwd <- as.vector(dbTable[1,1])

## Loading the BABS password ##
if (upload.results.to.database){
    FN <- paste0(hpc.mount, "Projects/reference_data/pwd_folder/babs.txt")
    dbTable <- read.delim(
      FN,
      header = F,
      sep = "\t",
      stringsAsFactors = F
    )
}

db.pwd <- as.vector(dbTable[1,1])
##                                                                           ##
###############################################################################

###################################
# Upload pca table to database    #
###################################

biologicSeqTools2::upload.pca.table.to.db(
    df.pca = Obio@dfPCA,
    host =Obio@dbDetailList$host,
    prim.data.db = Obio@dbDetailList$primDataDB,
    password = db.pwd,
    db.user = Obio@dbDetailList$db.user,
    PCAdbTableName = Obio@parameterList$PCAdbTableName
)

# Finished creating pca database table

###############################################################################
# Upload design file to database                                              #
###############################################################################
# Obio@parameterList$designdbTableName <- paste0(
#   Obio@parameterList$project_id, 
#     "_designTable"
# )
dfDesign4upload <- Obio@dfDesign
selVec <- c(
    "sampleID",
    "sample.id",
    "sample.group",
    "dataseries",
    names(dfDesign4upload)[grep("^f_",names(dfDesign4upload))],
    names(dfDesign4upload)[grep("^LRT_",names(dfDesign4upload))],
    names(dfDesign4upload)[grep("^comp_",names(dfDesign4upload))],
    names(dfDesign4upload)[grep("_color$",names(dfDesign4upload))]
)

pos <- grep("designTScol", names(Obio@parameterList))
if (length(pos) > 0 & !is.null(Obio@parameterList$designTScol)){
    selVec <- c(
        selVec,
        Obio@parameterList$designTScol
    )
}

selVec <- unique(selVec[selVec %in% names(dfDesign4upload)])

dfDesign4upload <- unique(dfDesign4upload[,selVec])
names(dfDesign4upload) <- gsub("\\.", "_", names(dfDesign4upload))

cmd.vec <- biologicSeqTools2::upload.datatable.to.database(
  host = Obio@dbDetailList$host,
  user = Obio@dbDetailList$db.user,
  password = db.pwd,
  prim.data.db = Obio@dbDetailList$primDataDB,
  dbTableName = Obio@parameterList$designdbTableName,
  df.data = dfDesign4upload,
  db.col.parameter.list = biologicSeqTools2::inferDBcategories(
      dfDesign4upload
  ),
    
  new.table = TRUE
)

killDbConnections()


###############################################################################
# Upload to database                                                          #
###############################################################################
Obio@databaseTable$Gene_description <- NULL

cmd.vec <- biologicSeqTools2::upload.datatable.to.database(
    host = Obio@dbDetailList$host,
    user = Obio@dbDetailList$db.user,
    password = db.pwd,
    prim.data.db = Obio@dbDetailList$primDataDB,
    dbTableName = Obio@parameterList$rnaseqdbTableName,
    # df.data = Obio@databaseTable[Obio@databaseTable$count_cut_off > 1, ], Use this option to filter reads
    df.data = Obio@databaseTable,
    # db.col.parameter.list = biologicSeqTools2::inferDBcategories(Obio@databaseTable[Obio@databaseTable$count_cut_off > 1, ]),
    db.col.parameter.list = biologicSeqTools2::inferDBcategories(Obio@databaseTable),
    new.table = TRUE
)

killDbConnections()


#setwd(Obio@parameterList$workdir)
doQuery <- function(
    Obio, 
    query,
    resOut = FALSE
){
    library(RMySQL)
    res <- NULL
    dbDB <- dbConnect(
        drv = RMySQL::MySQL(), 
        user = Obio@dbDetailList$db.user, 
        password = db.pwd, 
        host = Obio@dbDetailList$host,
        dbname = Obio@dbDetailList$primDataDB
    ) 
    tryCatch(res <- dbGetQuery(dbDB, query), error = function(c) {
        c$message <- stop(paste0("Error in ", query, "."))
    })
    dbDisconnect(dbDB)
    if (resOut){
        return(res)
    }
}


## Create user in db
query0 <- paste0("SELECT User, Host FROM mysql.user WHERE User = '",username,"' AND Host = '10.%';")
res <- doQuery(Obio, query = query0, resOut=TRUE)
if (nrow(res) == 0){
    query1 <- paste0(
        "CREATE USER '",
        username, 
        "'@'10.%' IDENTIFIED BY '",
        pass,
        "';"
    )
  
    doQuery(Obio, query = query1)
    #query0a <- paste0("DROP USER '",username,"'@'10.%';")   
    #doQuery(Obio, query = query0a)
} 


query2 <- paste0("GRANT SELECT on ",Obio@dbDetailList$primDataDB,".",designTB ," TO ",username,"@'10.%';")
doQuery(Obio, query = query2)

query3 <- paste0("GRANT SELECT on ",Obio@dbDetailList$primDataDB,".",mainTB ," TO ",username,"@'10.%';")
doQuery(Obio, query = query3)

query4 <- paste0("GRANT SELECT on ",Obio@dbDetailList$primDataDB,".",pcaTB ," TO ",username,"@'10.%';")
doQuery(Obio, query = query4)



###############################################################################
# Create microwebsite                                                         #
###############################################################################

## Default sample order: Order in the design file

sample.order <- unique(paste0(
  "norm_counts_",
  Obio@dfDesign$sample.id,
  "_TPM"
))




## Sort database.table ##
orderVec <- names(Obio@databaseTable)
lg2Order <- gsub("norm_counts_", "lg2_avg_", sample.order)

orderVec <- orderVec[!(orderVec %in% sample.order)]
orderVec <- orderVec[!(orderVec %in% lg2Order)]

orderVec <- c(
    orderVec,
    sample.order,
    lg2Order
)

Obio@databaseTable <- Obio@databaseTable[,orderVec]

## Get relevant colors ##
dfColor <- unique(Obio@dfDesign[,c("sample.id", "sample.group_color")])
row.names(dfColor) <- dfColor$sample.id

sample.colors <- sample.order
sample.colors <- gsub("_TPM", "", sample.colors)
sample.colors <- gsub("norm_counts_", "", sample.colors)

dfColor <- dfColor[sample.colors, ]

sample.colors <- as.vector(dfColor$sample.group_color)


if (!exists("gsea.cat.lines")){
    gsea.cat.lines <- ""
    downloadCatEnrichmentFNxlsx <- ""



} else {
    downloadCatEnrichmentFNxlsx <- paste0("outputs/", Obio@parameterList$project.code, ".enriched.categories.txt")

    gseaList <- list(
        "enriched_categories_table" = paste0(
            "p",
            substr(Obio@parameterList$project_id, 4,100),
            "_enriched_categories_table"
        ),
        "NES_array" = "array('NES_Skin_Card14KI_vs_Skin_WT');",
        "padj_array" = "array('padj_Skin_Card14KI_vs_Skin_WT');",
        "image_link_array" = "array('contrast_1_image_link');"
    )
}

if (!exists("timecourse.cat.lines")){
    timecourse.cat.lines <- NA
}

#library(SBwebtools)
setwd(Obio@parameterList$localWorkDir)

#library(SBwebtools)
webSiteDir <- Obio@parameterList$localWorkDir

project.code <- Obio@parameterList$project_id

setwd(Obio@parameterList$localWorkDir)

###############################################################################
## Deploy next generation website                                            ##



###############################################################################
## Do default Venn selection                                                 ##
defaultVennSel <- names(Obio@databaseTable)
defaultVennSel <- c(
    defaultVennSel[grep("contrast_", defaultVennSel)],
    defaultVennSel[grep("add_venn_", defaultVennSel)],
    defaultVennSel[grep("_TsCor_", defaultVennSel)]

)

rmVec <- c(
    grep("contrast_P_", defaultVennSel),
    grep("_lg2BaseMean_", defaultVennSel),
    grep("_padj_LRT_", defaultVennSel),
    grep("_lg10p_", defaultVennSel),
    grep("_stat_", defaultVennSel),
    grep("pValueCor_", defaultVennSel)
)

if (length(rmVec) > 0){
    defaultVennSel <- defaultVennSel[-rmVec]
}

## Add lrt
defaultVennSel <- c(
    defaultVennSel,
    names(Obio@databaseTable)[grep("contrast_L_lg10p", names(Obio@databaseTable))]
)

## Done with LRT selection                                                   ##
###############################################################################

###############################################################################
## Create default plot selection                                             ##
defaults <- sort(names(Obio@databaseTable)[grep("contrast_", names(Obio@databaseTable))])

xdefault <- defaults[grep("contrast_1_logFC", defaults)]
if (length(xdefault) == 0){
    xdefault <- defaults[1]
}

ydefault <- defaults[grep("contrast_1_lg10p", defaults)]
if (length(ydefault) == 0){
    ydefault <- defaults[2]
}



plotSelection <- c(
    names(Obio@databaseTable)[grep("_logFC_",names(Obio@databaseTable))],
    names(Obio@databaseTable)[grep("contrast_P_PCA_estimatePCA",names(Obio@databaseTable))],

    names(Obio@databaseTable)[grep("_lg10p_",names(Obio@databaseTable))],

    names(Obio@databaseTable)[grep("lg2BaseMean",names(Obio@databaseTable))],
    names(Obio@databaseTable)[grep("corCoef",names(Obio@databaseTable))]
)

chunk1 <- plotSelection[grep("_logFC_", plotSelection)]
chunkh2 <- plotSelection[grep("_lg10p_", plotSelection)]
chunk2a <-chunkh2[grep("_lg10p_PCA", chunkh2)]
chunk2b <- chunkh2[!(chunkh2  %in% chunk2a)]

chnk2c <- plotSelection[grep("_corCoef", plotSelection)]

chunk3 <- plotSelection[grep("_PCA_est", plotSelection)]

chunk4 <- plotSelection[grep("_lg2BaseMean_", plotSelection)]
# rmVec <- grep("_LRT_", chunk4)
# if (length(rmVec) > 0){
#     chunk4 <- chunk4[-rmVec]
# }

plotSelectionNew <- c(
    chunk1,
    chunk2b,
    chnk2c,
    chunk3,
    chunk2a,
    chunk4
)

plotAdd <- plotSelection[!(plotSelection %in% plotSelectionNew)]

plotSelection <- c(
    plotSelectionNew,
    plotAdd
)


## Done creating default plot selection                                      ##
###############################################################################

pos <- grep("designTScol", names(Obio@parameterList))
if (length(pos) == 1){
    timepoint <- Obio@parameterList$designTScol
} else {
    timepoint <- NULL
}

biologicSeqTools2::createSettingsFile(
    obj = Obio,
    df.data = Obio@databaseTable,
    defaultXcolName = xdefault,
    defaultYcolName = ydefault,
    timepointName = timepoint,
    primDataTable = Obio@parameterList$rnaseqdbTableName,
    pcaDbTable = Obio@parameterList$PCAdbTableName,
    sample.order = sample.order, #set to "" to go with default sorting
    heatmapSampleOrder = "",
    sample.names = "", # default is sample.order
    count.sample.colors = sample.colors,
    ptm.colum = "",
    count.table.headline = Obio@parameterList$count.table.headline,
    count.table.sidelabel = Obio@parameterList$count.table.sidelabel,
    venn.slider.selector.strings = defaultVennSel,
    plot.selection.strings = plotSelection,
    webSiteDir = paste0(hpc.mount, "Stefan/protocol_files/github/biologic/src/experiments/"),
    upper_heatmap_limit = 3,
    lower_heatmap_limit = -3,
    heamap.headline.text = Obio@parameterList$heamap.headline.text,
    project_id = Obio@parameterList$project_id
)

setwd(Obio@parameterList$localWorkDir)
## Done deploying next generation website                                    ##
###############################################################################

###############################################################################
## Create settings JSON                                                      ##
projectJSON <- biologicSeqTools2::createSettingsJSON(
    obj = Obio,
    df.data = Obio@databaseTable,
    defaultXcolName = xdefault,
    defaultYcolName = ydefault,
    timepointName = timepoint,
    primDataTable = Obio@parameterList$rnaseqdbTableName,
    pcaDbTable = Obio@parameterList$PCAdbTableName,
    sample.order = sample.order, #set to "" to go with default sorting
    heatmapSampleOrder = "",
    sample.names = "", # default is sample.order
    count.sample.colors = sample.colors,
    ptm.colum = "",
    count.table.headline = Obio@parameterList$count.table.headline,
    count.table.sidelabel = Obio@parameterList$count.table.sidelabel,
    venn.slider.selector.strings = defaultVennSel,
    plot.selection.strings = plotSelection,
    webSiteDir = paste0(hpc.mount, "Stefan/protocol_files/github/biologic/src/experiments/"),
    upper_heatmap_limit = 3,
    lower_heatmap_limit = -3,
    heamap.headline.text = Obio@parameterList$heamap.headline.text,
    project_id = Obio@parameterList$project_id
)

## Done                                                                      ##
###############################################################################

projectJSONstring <- gsub("[\n]", "", toString(projectJSON))
projectJSONstring <- gsub(" ", "", projectJSONstring)



###############################################################################
## Upload JSON

dfUpload <- data.frame(
    experiment_code = Obio@projectDetailList$project_id,
    experiment_settings = projectJSONstring
)

###############################################################################
## Check if an entry for this project exist already. If it does > update     ##

dbDB <- DBI::dbConnect(
    drv = RMySQL::MySQL(), 
    user = Obio@dbDetailList$db.user, 
    password = db.pwd, 
    host = Obio@dbDetailList$host, 
    dbname = "operational"
)

query <- paste0("SELECT id, experiment_code, created_at FROM settings WHERE experiment_code = '",Obio@parameterList$project_id,"';")

dfTest <- DBI::dbGetQuery(dbDB, query)
DBI::dbDisconnect(dbDB)

timeStamp <- format(as.POSIXlt(Sys.time()),'%Y-%m-%d %H:%M:%S')

if (nrow(dfTest) ==1){
    dfUpload[["id"]] = as.vector(dfTest[1,"id"])
    dfUpload[["created_at"]] = as.vector(dfTest[1,"created_at"])
    ## Delete original row ##
    dbDB <- DBI::dbConnect(
        drv = RMySQL::MySQL(), 
        user = Obio@parameterList$db.user, 
        password = db.pwd, 
        host = Obio@dbDetailList$host, 
        dbname = "operational"
    )
    query <- paste0("DELETE FROM settings WHERE experiment_code = '",Obio@parameterList$project_id,"';")
    res <- DBI::dbGetQuery(dbDB, query)
    DBI::dbDisconnect(dbDB)
    
    
} else {
    dfUpload[["created_at"]] = timeStamp
}

dfUpload[["updated_at"]] <- timeStamp
##                                                                           ##
###############################################################################

biologicSeqTools2::upload.datatable.to.database(
    host = Obio@dbDetailList$host,
    user = Obio@dbDetailList$db.user,
    password = db.pwd,
    prim.data.db = "operational",
    dbTableName = "settings",
    df.data = dfUpload,
    db.col.parameter.list = list(
        "INT(10) NULL DEFAULT NULL" = c("id"),
        "LONGTEXT NULL DEFAULT NULL" = c("^experiment_settings$"),
        "VARCHAR(255) CHARACTER SET latin1 COLLATE latin1_swedish_ci" = c("^experiment_code$"),
        "TIMESTAMP NULL DEFAULT CURRENT_TIMESTAMP" = c("created_at", "updated_at")
    ),
    increment = 5000,
    new.table = FALSE,
    first.row.name.index = 1,
    startOnlyWithConnectionCount1 = FALSE,
    cols2Index = NULL,
    indexName = NULL,
    mode = "MySQL",
    addRowNamesColumn = FALSE
)

## Done with uploading settings JSON                                         ##
###############################################################################

###############################################################################
## Add project                                                               ##

## project ##
timeStamp <- format(as.POSIXlt(Sys.time()),'%Y-%m-%d %H:%M:%S')

if (is.null(Obio@parameterList$projectName)){
    projectName <- paste0(Obio@parameterList$labname, " lab projects")
    Obio@parameterList$projectName <- projectName
}

## Check if project exists or needs to be generated ##
###############################################################################
## Check if an entry for this project exist already. If it does > update     ##

dbDB <- DBI::dbConnect(
    drv = RMySQL::MySQL(), 
    user = Obio@dbDetailList$db.user, 
    password = db.pwd, 
    host = Obio@dbDetailList$host, 
    dbname = "operational"
)

query <- paste0("SELECT * FROM projects WHERE name = '",Obio@projectDetailList$projectName,"';")

dfTest <- DBI::dbGetQuery(dbDB, query)
DBI::dbDisconnect(dbDB)

if (nrow(dfTest) == 0){
    ## Create new project ##
    myFun <- function(n = 5000) {
        a <- do.call(paste0, replicate(3, sample(LETTERS, n, TRUE), FALSE))
        paste0(a, sprintf("%04d", sample(9999, n, TRUE)), sample(LETTERS, n, TRUE))
    }
    
    if (is.null(Obio@parameterList$project_code)){
        project_code <- paste0("CRK",Obio@parameterList$labname, myFun(1))
        Obio@parameterList$project_code <- project_code
    }
    
    if (is.null(Obio@parameterList$project_description)){
        Obio@parameterList$project_description <- "Project Description TBD"
    }
    
    dfProject <-  data.frame(
        name = Obio@parameterList$projectName,
        project_code = Obio@parameterList$project_code,
        description = Obio@parameterList$project_description,
        lab = Obio@parameterList$labname
    )
    
    dfProject[["created_at"]] = timeStamp
    
    dfProject[["updated_at"]] <- timeStamp
    
    upload.datatable.to.database(
        host = Obio@dbDetailList$host,
        user = Obio@dbDetailList$db.user,
        password = db.pwd,
        prim.data.db = "operational",
        dbTableName = "projects",
        df.data = dfProject,
        db.col.parameter.list = list(
            "INT(10) NULL DEFAULT NULL" = c("id"),
            "LONGTEXT NULL DEFAULT NULL" = c("^description$"),
            "VARCHAR(255) CHARACTER SET latin1 COLLATE latin1_swedish_ci" = c("^name$", "^project_code$", "^lab$"),
            "TIMESTAMP NULL DEFAULT CURRENT_TIMESTAMP" = c("created_at", "updated_at")
        ),
        increment = 5000,
        new.table = FALSE,
        first.row.name.index = 1,
        startOnlyWithConnectionCount1 = FALSE,
        cols2Index = NULL,
        indexName = NULL,
        mode = "MySQL",
        addRowNamesColumn = FALSE
    )
    
} 


##                                                                           ##
###############################################################################

###############################################################################
## Add experiment                                                            ##

## Check if experiment exists already ##
timeStamp <- format(as.POSIXlt(Sys.time()),'%Y-%m-%d %H:%M:%S')

if (is.null(Obio@parameterList$project_id)){
    projectName <- paste0(Obio@parameterList$labname, " lab projects")
    Obio@parameterList$projectName <- projectName
}

## Check if project exists or needs to be generated ##
###############################################################################
## Check if an entry for this project exist already. If it does > update     ##

dbDB <- DBI::dbConnect(
    drv = RMySQL::MySQL(), 
    user = Obio@dbDetailList$db.user, 
    password = db.pwd, 
    host = Obio@dbDetailList$host, 
    dbname = "operational"
)

query <- paste0("SELECT * FROM experiments WHERE experiment_code = '",Obio@parameterList$project_id,"';")

dfTest <- DBI::dbGetQuery(dbDB, query)
DBI::dbDisconnect(dbDB)

if (nrow(dfTest) == 0){
   
    dfUpload <- data.frame(
        name = Obio@parameterList$projectName,
        project_id = Obio@parameterList$project_code,
        experiment_code = Obio@parameterList$project_id,
        main_question = Obio@parameterList$main_question,
        description = Obio@parameterList$description,
        lab = paste0(Obio@parameterList$labname, " Lab"),
        lead_scientist = 
    ) 
}


###############################################################################
## Upload gene categories creaed in this project                             ##

createProjectRefDbTables <- T

if (createProjectRefDbTables){
    ## Assign category IDs ##
    ## First check if categories exist already
    ## if they do - delete those and re-assign cat id
    ## if they don't find next available cat id
    for (i in 1:length(HmDisplayCatsFromDb)){
        dbDB <- DBI::dbConnect(
            drv = RMySQL::MySQL(), 
            user = Obio@dbDetailList$db.user, 
            password = db.pwd, 
            host = Obio@dbDetailList$host, 
            dbname = Obio@dbDetailList$ref.cat.db
        )
        
        query <- paste0("SELECT * FROM ",lab.categories.table," WHERE cat_name = '", names(HmDisplayCatsFromDb)[i], "' and cat_type = '", HmDisplayCatsFromDb[[i]]$cat_type, "';")
        dfOut <-  dbGetQuery(dbDB, query)
        dbDisconnect(dbDB)
        
        if (nrow(dfOut) == 1){
            HmDisplayCatsFromDb[[i]][["cat_id"]] <- dfOut$cat_id
        } else {
          HmDisplayCatsFromDb[[i]][["cat_id"]] <- NULL
        }
        
    }
    
    
    
    ## Delete all existing tables for this project ##
    dbDB <- DBI::dbConnect(
            drv = RMySQL::MySQL(), 
            user = Obio@dbDetailList$db.user, 
            password = db.pwd, 
            host = Obio@dbDetailList$host, 
            dbname = Obio@dbDetailList$ref.cat.db
    )
    
    ## Insert categories ##
    
    UploadList <- HmDisplayCatsFromDb
    
    ## Translate geneIDs to human for non-standard species ##
    rmVec <- as.vector(NULL, mode="numeric")
    
    if (geneIDcolumn != geneSelCol){
        for (d in 1:length(UploadList)){
            convertedGeneVec <- dfMainData[dfMainData[,geneIDcolumn] %in% UploadList[[d]]$geneVec ,geneSelCol]
            convertedGeneVec <- unique(convertedGeneVec)
            convertedGeneVec <- convertedGeneVec[convertedGeneVec != ""]
            convertedGeneVec <- convertedGeneVec[convertedGeneVec != 0]
            
            if (length(convertedGeneVec > 0)){
                UploadList[[d]]$geneVec <- convertedGeneVec
            } else {
                rmVec <- c(
                  rmVec, 
                  d
                )
            }
            
        }
    }
    
    if (length(rmVec) > 0){
        UploadList <- UploadList[-rmVec]
    }
    
        
    ## Upload to database ##
    createProjectRefDbTables <- T
    if (createProjectRefDbTables){  
        for (i in 1:length(UploadList)){
            ###########################################################################
        ## check if cat name exist and delete if it does                         ##
        #library(RMySQL)
        dbDB <- DBI::dbConnect(
            drv = RMySQL::MySQL(), 
            user = Obio@dbDetailList$db.user, 
            password = db.pwd, 
            host = Obio@dbDetailList$host, 
            dbname = Obio@dbDetailList$ref.cat.db
        )
        
        query <- paste0("SELECT cat_id, cat_name FROM ",lab.categories.table, " WHERE cat_name = '",names(UploadList)[i],"'")
        
        dfTest <- dbGetQuery(dbDB, query)
        dbDisconnect(dbDB)
        ## Done                                                                  ##
        ###########################################################################
        if (nrow(dfTest) ==1){
            catID = as.vector(dfTest[1,"cat_id"])
        } else {
        
            catID <- biologicSeqTools2::add.category.to.lab.reference.table.hs(
            host = Obio@dbDetailList$host,
            pwd = db.pwd,
            user = Obio@dbDetailList$db.user,
            cat.ref.db = "reference_categories_db_new",
            cat.ref.db.table = lab.categories.table,
            gene.vector = UploadList[[i]]$geneVec,
            gene.id = geneSelCol, #options hgnc_symbol, mgi_symbol
            mm.hs.conversion.file =  paste0(hpc.mount, "Projects/reference_data/20160303.homologene.data.txt"),
            cat_name = names(UploadList)[i],
            cat_type = UploadList[[i]]$cat_type,
            data_source = UploadList[[i]]$data_source,
            comments_1 = "",
            comments_2 = "",
            new.lab.category.table = F,
            cat.description.db  = "internal_categories",
            cat.description.db.table = "category_description",
            cat.description.text = UploadList[[i]]$cat.description.text,
            lab.name = labname
        )
        }
        
        UploadList[[i]][["cat_id"]] <- catID
        
        ## Add cat id to main list
        HmDisplayCatsFromDb[[names(UploadList)[i]]][["cat_id"]] <- catID
      
        }
    }
}


## Done                                                                      ##
###############################################################################


```

