<!-- Set PArameters Module -->
<!-- Set the chnkPrefix to make all chunks unique in the overall folder -->
```{r partC_C8_hm_init, echo=FALSE, eval=TRUE, warning=FALSE}
chnkPrefix <- "C8.Heatmap."
VersionPdfExt <- VersionPdfExt <- paste0(".",chnkPrefix,"V", gsub("-", "", Sys.Date()), ".pdf")

```



```{r Heatmaps_data_prep, echo=TRUE, eval=TRUE, warning=FALSE, results=F}
## Heatmap setup ##
## Select Heatmap samples ##
FN <- paste0(hpc.mount, "Projects/reference_data/pwd_folder/babs.txt")
dbTable <- read.delim(
  FN,
  header = F,
  sep = "\t",
  stringsAsFactors = F
)

db.pwd <- as.vector(dbTable[1,1])


selVec <- c(
    Obio@parameterList$geneIDcolumn,
    names(Obio@databaseTable)[grep("norm_counts", names(Obio@databaseTable))]
)

chnkVec <- as.vector(NULL, mode = "character")
HMplotList <- list()

if (Obio@parameterList$geneIDcolumn == "mgi_symbol" | Obio@parameterList$geneIDcolumn == "hgnc_symbol"){
    geneSelCol <- Obio@parameterList$geneIDcolumn
} else {
    geneSelCol <- "hgnc_symbol"
}

if (is.null(Obio@parameterList$HmDisplayCatsFromDb)){
    Obio@parameterList[["HmDisplayCatsFromDb"]] <- list()
}

###############################################################################
## Get gene sets to retrieve                                                 ##


## Start with Nmost variable genes ##
if (!is.null(Obio@dataTableList$Ntop4pcaGeneSelection) | 
    length(Obio@dataTableList$Ntop4pcaGeneSelection) > 3){

    dfDataTable <- Obio@databaseTable
    
    geneVec <- as.vector(unique(dfDataTable[dfDataTable[,Obio@parameterList$primaryAlignmentGeneID] %in% Obio@dataTableList$Ntop4pcaGeneSelection,geneSelCol]))
}

## Insert gene set into database ##
createProjectRefDbTables <- T

if (createProjectRefDbTables){    
    ## Delete all previous entries ##
    #library(RMySQL)
    dbDB <- DBI::dbConnect(
        drv = RMySQL::MySQL(), 
        user = Obio@dbDetailList$db.user, 
        password = db.pwd, 
        host = Obio@dbDetailList$host, 
        dbname = Obio@dbDetailList$ref.cat.db
    )
    
    cat_type <- paste0("temp_", Obio@parameterList$project_id)
    query <- paste0("DELETE FROM ",Obio@parameterList$lab.categories.table," WHERE cat_type = '", cat_type, "'")
    dfOut <-  dbGetQuery(dbDB, query)
    dbDisconnect(dbDB)
    
    
    
    cat.name <- paste0("Experiment_",Obio@parameterList$project_id, "_",Obio@parameterList$NtopGene,"_most_variable_genes")
    cat_type <- paste0("temp_", Obio@parameterList$project_id)
    data_source <- paste0(Obio@parameterList$labname, " Lab") 
    cat.description.text <- paste0("In this gene set the ",Obio@parameterList$NtopGene," most variable genes from ",Obio@parameterList$labname," lab experiment \\<a href=\\'https:\\/\\/biologic.crick.ac.uk\\/",Obio@parameterList$project_id,"\\'\\>",Obio@parameterList$project_id, "\\<\\/a\\> are compiled.")
    
    ref.cat.db.table = Obio@parameterList$lab.categories.table
    
    ###########################################################################
    ## check if cat name exist and delete if it does                         ##
    #library(RMySQL)
    dbDB <- DBI::dbConnect(
        drv = RMySQL::MySQL(), 
        user = Obio@dbDetailList$db.user, 
        password = db.pwd, 
        host = Obio@dbDetailList$host, 
        dbname = Obio@dbDetailList$ref.cat.db
    )
    
    query <- paste0("SELECT cat_id, cat_name FROM ",Obio@parameterList$lab.categories.table, " WHERE cat_name = '",cat.name,"'")
    
    dfTest <- dbGetQuery(dbDB, query)
    dbDisconnect(dbDB)
    ## Done                                                                  ##
    ###########################################################################
    if (nrow(dfTest) ==1){
        catID = as.vector(dfTest[1,"cat_id"])
    } else {
    
        catID <- add.category.to.lab.reference.table.hs(
        host = Obio@dbDetailList$host,
        pwd = db.pwd,
        user = Obio@dbDetailList$db.user,
        cat.ref.db = "reference_categories_db_new",
        cat.ref.db.table = Obio@parameterList$lab.categories.table,
        gene.vector = geneVec,
        gene.id = geneSelCol, #options hgnc_symbol, mgi_symbol
        mm.hs.conversion.file =  paste0(hpc.mount, "Projects/reference_data/20160303.homologene.data.txt"),
        cat_name = cat.name,
        cat_type = cat_type,
        data_source = paste0(Obio@parameterList$labname, " lab experiment ", Obio@parameterList$project_id, "."),
        comments_1 = "",
        comments_2 = "",
        new.lab.category.table = F,
        cat.description.db  = "internal_categories",
        cat.description.db.table = "category_description",
        cat.description.text = cat.description.text,
        lab.name = Obio@parameterList$labname
    )
    }
    
    ## Enter into heatmap table ##
    Obio@parameterList$HmDisplayCatsFromDb[[cat.name]] <- catID
    
    
    ###########################################################################
    ## Create gene selections for each logFC comparison                      ##
    logFCselections <- names(Obio@databaseTable)[grep("_logFC_", names(Obio@databaseTable))]
    padjSelections <- gsub("_logFC_", "_padj_", logFCselections)
    
    dfSelections <- data.frame(logFCselections, padjSelections)
    dfSelections <- dfSelections[dfSelections[,"padjSelections"] %in% names(Obio@databaseTable),]
    
    for (k in 1:nrow(dfSelections)){
        dfDataTable <- Obio@databaseTable
        
        padjCutOff <- 0.05
        
        geneVec <- as.vector(
            unique(
                dfDataTable[dfDataTable[,as.vector(dfSelections$padjSelections[k])] < 0.05 & dfDataTable[,as.vector(dfSelections$logFCselections[k])] != 0,geneSelCol]
            )
        )
        
        if (length(geneVec) > 1500){
            padjCutOff <- 0.01
        
            geneVec <- as.vector(
                unique(
                    dfDataTable[dfDataTable[,as.vector(dfSelections$padjSelections[k])] < 0.05 & 
                                    dfDataTable[,as.vector(dfSelections$logFCselections[k])] != 0,geneSelCol
                    ]
                )
            )
        }
        
        ## Insert gene set into database ##
            cat.name <- paste0(
                "Experiment_",Obio@parameterList$project_id, "_",dfSelections$padjSelections[k],"_smaller_than_", gsub("[.]", "_", padjCutOff))
            cat_type <- paste0("temp_", Obio@parameterList$project_id)
            data_source <- paste0(Obio@parameterList$labname, " Lab") 
            cat.description.text <- paste0(
                "In this gene set the genes that exhibited an adjusted p value of less than ", 
                padjCutOff, 
                " in the differential gene expression comparsion ", 
                as.vector(dfSelections$logFCselections[k]),
                " in ",
                Obio@parameterList$labname,
                " lab experiment \\<a href=\\'https:\\/\\/biologic.crick.ac.uk\\/",Obio@parameterList$project_id,"\\'\\>",Obio@parameterList$project_id, "\\<\\/a\\> are compiled."
            )
            
            ref.cat.db.table = Obio@parameterList$lab.categories.table
            
            
             ###########################################################################
    ## check if cat name exist and delete if it does                         ##
    #library(RMySQL)
    dbDB <- DBI::dbConnect(
        drv = RMySQL::MySQL(), 
        user = Obio@dbDetailList$db.user, 
        password = db.pwd, 
        host = Obio@dbDetailList$host, 
        dbname = Obio@dbDetailList$ref.cat.db
    )
    
    query <- paste0("SELECT cat_id, cat_name FROM ",Obio@parameterList$lab.categories.table, " WHERE cat_name = '",cat.name,"'")
    
    dfTest <- dbGetQuery(dbDB, query)
    dbDisconnect(dbDB)
    ## Done                                                                  ##
    ###########################################################################
    if (nrow(dfTest) ==1){
        catID = as.vector(dfTest[1,"cat_id"])
    } else {
            catID <- add.category.to.lab.reference.table.hs(
            host = Obio@dbDetailList$host,
            pwd = db.pwd,
            user = Obio@dbDetailList$db.user,
            cat.ref.db = "reference_categories_db_new",
            cat.ref.db.table = Obio@parameterList$lab.categories.table,
            gene.vector = geneVec,
            gene.id = geneSelCol, #options hgnc_symbol, mgi_symbol
            mm.hs.conversion.file =  paste0(hpc.mount, "Projects/reference_data/20160303.homologene.data.txt"),
            cat_name = cat.name,
            cat_type = cat_type,
            data_source = paste0(Obio@parameterList$labname, " lab experiment ", Obio@parameterList$project_id, "."),
            comments_1 = "",
            comments_2 = "",
            new.lab.category.table = FALSE,
            cat.description.db  = "internal_categories",
            cat.description.db.table = "category_description",
            cat.description.text = cat.description.text,
            lab.name = Obio@parameterList$labname
        )
            
    }
            ## Enter into heatmap table ##
            Obio@parameterList$HmDisplayCatsFromDb[[cat.name]] <- catID
                
            
            
            ##                                                                       ##
            ###########################################################################
        
    }


    }


###############################################################################
##                                                                           ##
# #Define cat id tables manually
#  Obio@parameterList$HmDisplayCatsFromDb[["Experiment_csl325_500_most_variable_genes"]] <- "cs_lab_categories__10"
# 
# Obio@parameterList$HmDisplayCatsFromDb[["Experiment_csl325_contrast_1_padj_M_Zfp516NULL_vs_M_WT_smaller_than_0_05"]] <- "cs_lab_categories__11"
# 
# Obio@parameterList$HmDisplayCatsFromDb[["Experiment_csl325_contrast_2_padj_WT_Female_vs_WT_Male_smaller_than_0_05"]] <- "cs_lab_categories__12"


##                                                                           ##
###############################################################################

###############################################################################
## Make Heatmaps for all gene sets in HmList                                 ##

###############################################################################
## Reorder Obio@parameterList$HmDisplayCatsFromDb so that 500 var is on top  ##

pos <- grep("most_variable_genes", names(Obio@parameterList$HmDisplayCatsFromDb))

if (length(pos) > 0){
  pos <- pos[1]
  newOrder <- c(
    names(Obio@parameterList$HmDisplayCatsFromDb)[pos],
    names(Obio@parameterList$HmDisplayCatsFromDb)[-pos]
  )
  Obio@parameterList$HmDisplayCatsFromDb <- Obio@parameterList$HmDisplayCatsFromDb[newOrder]
  
  
  
}

##                                                                           ##
###############################################################################

## Begin heatmap plotting loop ##
for (k in 1:length(Obio@parameterList$HmDisplayCatsFromDb)){
    
    geneSel <- retrieve.gene.category.from.db(
        cat_id = Obio@parameterList$HmDisplayCatsFromDb[[k]], 
        gene.symbol=geneSelCol, 
        password=db.pwd, 
        host=Obio@dbDetailList$host, 
        user=Obio@dbDetailList$db.user
    )
    
    geneSel <- unique(geneSel)
    
    if (length(geneSel) > 2){
        dfDataTable <- Obio@databaseTable
        dfDataTable <- unique(dfDataTable[dfDataTable[, geneSelCol] %in% geneSel, selVec])
        
        dfHmBase <- unique(dfDataTable[,selVec])
        
        while (sum(duplicated(dfHmBase[, Obio@parameterList$geneIDcolumn])) > 0){
            dfHmBase[duplicated(dfHmBase[, Obio@parameterList$geneIDcolumn]), Obio@parameterList$geneIDcolumn] <- paste0(
                dfHmBase[duplicated(dfHmBase[, Obio@parameterList$geneIDcolumn]), 
                Obio@parameterList$geneIDcolumn], "_", i
            )
            i=i+1
        }
        
        row.names(dfHmBase) <- dfHmBase[, Obio@parameterList$geneIDcolumn]

        dfHmBase[, Obio@parameterList$geneIDcolumn] <- NULL
        
        ## calculate row-means ##
        rowMeans <- apply(
            dfHmBase,
            1,
            function(x) mean(x)
        )
            
        rowMeans[rowMeans ==0] <- 0.001
            
        hmMax <- 4
        for (i in 1:ncol(dfHmBase)){
            dfHmBase[,i] <- log2(dfHmBase[,i] / rowMeans)
        }
            
        dfHmBase[dfHmBase > hmMax] <- hmMax
        dfHmBase[dfHmBase < -1*hmMax] <- -1*hmMax
            
            
        names(dfHmBase) <- gsub("norm_counts_", "", names(dfHmBase))
        names(dfHmBase) <- gsub("_TPM", "", names(dfHmBase))
            
        mHmBase <- data.matrix(dfHmBase)
            
        if ( nrow(mHmBase) < 51){
            showRowNames <- TRUE
        } else {
            showRowNames <- FALSE
        }
        
        ## Create heatmap plot ##
        library(ComplexHeatmap)
        library(circlize)
        f1 = colorRamp2(seq(-4, 4, length = 3), c("#3060cf", "#fffbbc","#c4463a"))    
    

        ## Create top annotation and colorbars ##
        # from https://www.biostars.org/p/368265/
        # from https://bioconductor.statistik.tu-dortmund.de/packages/3.1/bioc/vignettes/ComplexHeatmap/inst/doc/ComplexHeatmap.html
        
        
        anno <- as.data.frame(colnames(mHmBase))
        colnames(anno) <- "Sample"
        anno$Group <- sapply(as.vector(anno[,1]), function(x) paste0(unlist(strsplit(x, "_"))[1], "_",unlist(strsplit(x, "_"))[2]))
        
        ## Color sample groups in line with the designated sample group color ##
        
        #library(scales)
        #hue_pal()(2)
        df <- unique(data.frame(Obio@dfDesign[,c("sample.group", "sample.group_color")]))
        
        GroupVec <- as.vector(unique(df$sample.group_color))
        names(GroupVec) <- as.vector(unique(df$sample.group))
        
        df2 <- unique(data.frame(Obio@dfDesign[,c("sample.id","sample.group", "sample.group_color")]))
        df2 <- data.frame(df2[,c("sample.group")])
        names(df2) <- "Group"
        
        
        ha = HeatmapAnnotation(df = df2, col = list(Group = GroupVec))
    
        ht_opt(
            legend_border = "black",
            heatmap_border = TRUE,
            annotation_border = TRUE
        )
        
        HMplotList[[names(Obio@parameterList$HmDisplayCatsFromDb)[k]]] = Heatmap(
            mHmBase,
            column_title = gsub(
                    "_", 
                    " ", 
                    paste0("Heatmap_", names(Obio@parameterList$HmDisplayCatsFromDb)[k])
            ),
            name = paste0("HM_", k), 
            #row_km = 5,
            col = f1,
           
            show_column_names = T,
            show_row_names = showRowNames,
            border = TRUE,
            
            #Dendrogram configurations: columns
            clustering_distance_columns="euclidean",
            clustering_method_columns="complete",
            column_dend_height=unit(10,"mm"),
            
            #Dendrogram configurations: rows
            clustering_distance_rows="euclidean",
            clustering_method_rows="complete",
            row_dend_width=unit(10,"mm"),
            #top_annotation = ha,
            show_heatmap_legend = TRUE
            #row_title = NULL,
            #show_row_dend = FALSE
        ) 
        
    ht_opt(RESET = TRUE)
        
    link <- paste0("https://biologic.crick.ac.uk/",Obio@parameterList$project_id,"/category-view/",Obio@parameterList$HmDisplayCatsFromDb[[k]])
    
    ###########################################################################
    ## Save plot to file                                                     ##
    FNbase <- paste0("Heatmap.", names(Obio@parameterList$HmDisplayCatsFromDb)[k],VersionPdfExt)
    FN <- paste0(Obio@parameterList$reportFigDir, FNbase)
    FNrel <- paste0("report_figures/", FNbase)
    
    pdf(FN)
        print(HMplotList[[names(Obio@parameterList$HmDisplayCatsFromDb)[k]]])
    dev.off()
    ##                                                                       ##
    ###########################################################################
    
    figCap <- paste0(
    '**Figure ',
    figureCount,
    ':** Heatmap showing the gene category ', gsub('_', ' ', names(Obio@parameterList$HmDisplayCatsFromDb)[k]), '. ',
        'Download a pdf of this figure <a href="',FNrel,'" target="_blank">here</a>. ',
        'An interactive version of this heatmap with an option for further filtering can be found <a href="',link,'" target="_blank">here</a>.'
    )
    
    figureCount <- figureCount + 1 
    
    NewChnk <- paste0(
            "### HM_", names(Obio@parameterList$HmDisplayCatsFromDb)[k],
            "\n```{r Heatmap_HM_",k,", results='asis', echo=F, eval=TRUE, warning=FALSE, fig.cap='",figCap,"'}\n",
            "\n",
            "\n print(HMplotList[['",names(Obio@parameterList$HmDisplayCatsFromDb)[k],"']])",
            "\n cat(  '\n')",
            "\n\n\n```\n"   
    )
    
    chnkVec <- c(
        chnkVec,
        NewChnk
    )
    
    } ## End making heatmap 
    
## Done making heatmaps                                                      ##
###############################################################################
}
## End heatmap plotting loop

## Done                                                                      ##
###############################################################################


if (length(HMplotList) > 2){
    tabVar <- ".tabset .tabset-fade .tabset-dropdown"
} else {
    tabVar <- ".tabset .tabset-fade .tabset-pills"
}

```

## Gene Expression Overview {`r tabVar`}

```{r Heatmap_plot, echo=TRUE, eval=TRUE, warning=FALSE, results="asis"}
cat(paste(knitr::knit(text = chnkVec, quiet = T), collapse = '\n'))
```

