<!-- Set Parameters Module -->
<!-- Set the chnkPrefix to make all chunks unique in the overall folder -->
```{r, echo=TRUE, eval=TRUE, warning=FALSE}
chnkPrefix <- "edit.RSEM.tpm."
VersionPdfExt <- VersionPdfExt <- paste0(".",chnkPrefix,"V", gsub("-", "", Sys.Date()), ".pdf")

```

## GSEA
### Process GSEA
```{r, echo=TRUE, eval=TRUE, warning=FALSE, results="asis"}
# ###############################################################################
# ## Process GSEA outputs                                                      ##
#
# # # Subfunction to select relevant tables

dbDB = DBI::dbConnect(
    drv = RMySQL::MySQL(),
    user = Obio@dbDetailList$db.user,
    password = db.pwd,
    dbname = "reference_categories_db_new",
    host = Obio@dbDetailList$host
)

#This will move to the header


availableTables = dbGetQuery(dbDB, "SHOW TABLES in reference_categories_db_new")
availableTables <- as.vector(availableTables$Tables_in_reference_categories_db_new)
dbDisconnect(dbDB)
#
# ## Check if all reference tables are present in the db ##
if (!is.null(Obio@parameterList$GSEAtables)){
    tables <- Obio@parameterList$GSEAtables
} else {
    tables <- c(
        "mysigdb_h_hallmarks",
        "mysigdb_c5_BP" #,
        #Obio@parameterList$lab.categories.table
    )
}

print(
    paste0(
        sum(!(tables %in% availableTables)),
        " tables NOT present in the reference database."
    )
)
#
#
#
# #temporary change
# #tables <- tables[c(1,2,3,5,6)]
# #tables <- append(tables, "vt_lab_categories")
# # End of parts to be moved to the header
#
database.table2 <- Obio@databaseTable
rmVec <- c(
    #grep("contrast_2", names(database.table2))
    grep("contrast_P", names(database.table2)),
    grep("contrast_L", names(database.table2)),
    grep("LRT_", names(database.table2)),
    grep("PCA_", names(database.table2)),
    grep("norm_counts_", names(database.table2)),
    grep("intercept_", names(database.table2)),
    grep("r2_P", names(database.table2)),
    grep("DEseq2NormalizedReadCounts", names(database.table2)),
    grep("p_value_P", names(database.table2)),
    grep("lg2_avg_", names(database.table2))
)

if (length(rmVec) > 0){
    database.table2 <- database.table2[,-rmVec]
}



gsea.cat.lines <- biologicSeqTools::create.GSEA.table(
    GSEADir = paste0(Obio@parameterList$localWorkDir, "GSEA"),
    logFC.column.name = "logFC",
    host = Obio@dbDetailList$host,
    refdbname= "reference_categories_db_new",
    refDBTableName = Obio@parameterList$enriched.categories.dbTableName,
    db.user = Obio@dbDetailList$db.user,
    db.password = db.pwd,
    tables = tables,
    df.dataTable = database.table2,
    outputDir = Obio@parameterList$outputDir
)

## Done processing GSEA outputs                                              ##
###############################################################################


###############################################################################
## Optional Module: Correlation analysis                                     ##

## If required, run the module
# templateRNAseqAnalysis_pluginModule_correlationAnalysis.R
# here.

## DoneOptional Module: Correlation analysis                                 ##
###############################################################################


###############################################################################
## Select tables for reference listing                                       ##

## Switch to mgi symbols ##


## List all available reference tables ##
# list.db.tables.in.db(
#     dbname = "reference_categories_db_new",
#     password = db.pwd,
#     host=Obio@dbDetailList$host,
#     user=Obio@dbDetailList$db.user
# )

labCatName <-  paste0(Obio@parameterList$labname, " Lab")

if (length(Obio@referenceTableList) == 0){
    Obio@referenceTableList <- list(
        "Hallmark Signatures" = "mysigdb_h_hallmarks",
        "Pathways" = "mysigdb_c2_1329_canonical_pathways",
        "GO-BP" = "mysigdb_c5_BP",
        "GO-MF" = "mysigdb_c5_MF",
        "TF Motifs" = "TRANSFAC_and_JASPAR_PWMs",
        "Protein Complexes" = "networkcategories",
        "GO-BP" = "GO_Biological_Process_2017",
        "mysigdb_c7_immunologic_signatures" = "Immunological Signatures"
    )
}



Obio@referenceTableList[[labCatName]] <- Obio@parameterList$lab.categories.table

## Define relevant genes for selection ##
relevant.genes <- as.vector(
    unique(
        Obio@databaseTable[Obio@databaseTable$cluster_order, Obio@parameterList$geneIDcolumn]
    )
)

length(relevant.genes)

for (i in 1:length(Obio@referenceTableList)) {
    df.ref <- biologicSeqTools::import.db.table.from.db(
        host = Obio@dbDetailList$host,
        dbname = Obio@dbDetailList$ref.cat.db,
        dbtable = Obio@referenceTableList[[i]],
        user = Obio@dbDetailList$db.user,
        password = db.pwd
    )

    ## Remove temp categories ##
    # temPos <- grep("temp_", df.ref$cat_type)
    #
    # if (length(temPos) > 0) {
    #     df.ref <- df.ref[-temPos,]
    # }

    if (!(Obio@parameterList$geneIDcolumn %in% c("hgnc_symbol", "mgi_symbol"))){
        geneCol <- "hgnc_symbol"
    } else {
        geneCol <- Obio@parameterList$geneIDcolumn
    }
    
    df.temp <- biologicSeqTools::add2labCatSelectionDBtable(
        df.ref = df.ref,
        cat_group_name = names(Obio@referenceTableList)[i],
        reference.gene.vector = relevant.genes,
        ref.gene.vec.id = geneCol,
        cat_views = NA
    )

    if (i == 1) {
        df.db.table <- df.temp
    } else {
        df.db.table <- rbind(df.temp, df.db.table)
    }
}

###############################################################################
## Filter cat dataset                                                        ##
###############################################################################
## Remove all datasets with less than 5 proteins matched
catIdGroupsExemptFromFiltering <- c(
    Obio@parameterList$lab.categories.table,
    "mysigdb_h_hallmarks",
    "mysigdb_c2_1329_canonical_pathways",
    "mysigdb_c5_BP"
)

dfUnfiltered <- df.db.table[as.vector(unlist(sapply(catIdGroupsExemptFromFiltering, function(x) grep(x, df.db.table$cat_id)))),]

dim(df.db.table)
df.db.table <- df.db.table[df.db.table$cat_count >= 5,]
dim(df.db.table)

## Add unfiltered categories ##
if (exists("dfUnfiltered") & (nrow(dfUnfiltered) > 0)){
    df.db.table <- unique(
        rbind(
            dfUnfiltered,
            df.db.table
        )
    )
}

dim(df.db.table)

#df.db.table <- df.db.table[df.db.table$cat_weight > 0.3,]
#dim(df.db.table)


###############################################################################
## Add full temp categories associated with this project                     ##
## Lab categories ##
df.ref <- biologicSeqTools::import.db.table.from.db(
    dbname = Obio@dbDetailList$ref.cat.db,
    dbtable = Obio@parameterList$lab.categories.table,
    password = db.pwd,
    host = Obio@dbDetailList$host,
    user = Obio@dbDetailList$db.user
)

## Remove all project temp categories from df.ref
df.ref <- df.ref[grep(paste0("temp_", Obio@parameterList$project_id), df.ref$cat_type),]

if (exists("df.ref") & nrow(df.ref) > 0){
    df.temp <- add2labCatSelectionDBtable(
        df.ref = df.ref,
        cat_group_name = "This project",
        reference.gene.vector = reference.gene.vector,
        ref.gene.vec.id = Obio@parameterList$geneIDcolumn,
        cat_views = NA
    )

    df.db.table <- rbind(df.temp, df.db.table)
}
## Done adding temp categories associated with this project
###############################################################################


###############################################################################
## Upload to database                                                        ##
###############################################################################
biologicSeqTools::upload.datatable.to.database(
    host = Obio@dbDetailList$host,
    user = Obio@dbDetailList$db.user,
    password = db.pwd,
    prim.data.db = Obio@parameterList$cat.ref.db,
    dbTableName = Obio@parameterList$cat.ref.db.table,
    df.data = df.db.table,
    db.col.parameter.list = biologicSeqTools::inferDBcategories(df.db.table),
    new.table = TRUE
)

##  End creating reference category selection for this project               ##
###############################################################################

###############################################################################
# Create timecourse specifications                                            #
###############################################################################
# Add dataseries_color to dfDesign

#timecourse.cat.lines <- create.timecourse.cat.lines(
#     dfDesign
# )


###############################################################################
# Create microwebsite                                                         #
###############################################################################

## Color samples by sample group ##
# Get sample order from database.table
## oder samples according to PC1
# df.pca <- df.pca[order(df.pca$pca1, decreasing = FALSE),]
# sample.order <- paste0("norm_counts_", df.pca$sample_id)

if (!is.null(Obio@parameterList$designTScol)){
    dfOrder <- unique(Obio@dfDesign[,c("sample.id", "sample.group", Obio@parameterList$designTScol)])
    dfOrder <- dfOrder[order(dfOrder[,Obio@parameterList$designTScol], decreasing = F),]
} else {
    dfOrder <- (Obio@dfDesign[,c("sample.id", "sample.group")])
}

sample.order <- paste0("norm_counts_", dfOrder$sample.id, "_TPM")


#paste0("sample.order <- c('",paste(sample.order, collapse = "','"),"')")



## In case samples need to be reordered ##
#sample.order <- c('",paste(sample.order, collapse = "','"),"')")


#sample.order <- c('norm_counts_CTRL_D0_1_TPM','norm_counts_CTRL_D0_2_TPM','norm_counts_CTRL_D0_3_TPM','norm_counts_CTRL_D1_1_TPM','norm_counts_CTRL_D1_2_TPM','norm_counts_CTRL_D1_3_TPM','norm_counts_CTRL_D2_1_TPM','norm_counts_CTRL_D2_2_TPM','norm_counts_CTRL_D2_3_TPM','norm_counts_CTRL_D3_1_TPM','norm_counts_CTRL_D3_2_TPM','norm_counts_CTRL_D3_3_TPM','norm_counts_CTRL_D4_1_TPM','norm_counts_CTRL_D4_2_TPM','norm_counts_CTRL_D4_3_TPM','norm_counts_CTRL_D5_1_TPM','norm_counts_CTRL_D5_2_TPM','norm_counts_CTRL_D5_3_TPM','norm_counts_CTRL_D5_4_TPM','norm_counts_CTRL_D5_5_TPM','norm_counts_CTRL_D5_6_TPM','norm_counts_KLF17_D0_2_TPM','norm_counts_KLF17_D0_3_TPM','norm_counts_KLF17_D1_1_TPM','norm_counts_KLF17_D1_2_TPM','norm_counts_KLF17_D1_3_TPM','norm_counts_KLF17_D2_1_TPM','norm_counts_KLF17_D2_2_TPM','norm_counts_KLF17_D2_3_TPM','norm_counts_KLF17_D3_1_TPM','norm_counts_KLF17_D3_2_TPM','norm_counts_KLF17_D3_3_TPM','norm_counts_KLF17_D4_1_TPM','norm_counts_KLF17_D4_2_TPM','norm_counts_KLF17_D4_3_TPM','norm_counts_KLF17_D5_1_TPM','norm_counts_KLF17_D5_2_TPM','norm_counts_KLF17_D5_3_TPM'
# sample.order <- c(
#     'norm_counts_DEN_Fasted_1_TPM',
#     'norm_counts_DEN_Fasted_2_TPM',
#     'norm_counts_DEN_Fasted_3_TPM',
#     'norm_counts_DEN_Fed_1_TPM',
#     'norm_counts_DEN_Fed_2_TPM',
#     'norm_counts_DEN_Fed_3_TPM',
#     'norm_counts_nonDEN_Fasted_1_TPM',
#     'norm_counts_nonDEN_Fasted_2_TPM',
#     'norm_counts_nonDEN_Fasted_3_TPM',
#     'norm_counts_nonDEN_Fasted_4_TPM',
#     'norm_counts_nonDEN_Fed_1_TPM',
#     'norm_counts_nonDEN_Fed_2_TPM',
#     'norm_counts_nonDEN_Fed_3_TPM'
# )



# sample.order <- c(
#     'norm_counts_Glia_HpolyHet_1_TPM',
#     'norm_counts_Glia_HpolyHet_2_TPM',
#     'norm_counts_Glia_HpolyHet_3_TPM',
#     'norm_counts_Glia_HpolyHet_4_TPM',
#     'norm_counts_Glia_HpolyHom_1_TPM',
#     'norm_counts_Glia_HpolyHom_2_TPM',
#     'norm_counts_Glia_HpolyHom_3_TPM',
#     'norm_counts_Glia_HpolyHom_4_TPM',
#     'norm_counts_NonGlia_HpolyHet_1_TPM',
#     'norm_counts_NonGlia_HpolyHet_2_TPM',
#     'norm_counts_NonGlia_HpolyHet_3_TPM',
#     'norm_counts_NonGlia_HpolyHet_4_TPM',
#     'norm_counts_NonGlia_HpolyHom_1_TPM',
#     'norm_counts_NonGlia_HpolyHom_2_TPM',
#     'norm_counts_NonGlia_HpolyHom_3_TPM',
#     'norm_counts_NonGlia_HpolyHom_4_TPM'
# )

## Sort database.table ##
orderVec <- names(Obio@databaseTable)
lg2Order <- gsub("norm_counts_", "lg2_avg_", sample.order)

orderVec <- orderVec[!(orderVec %in% sample.order)]
orderVec <- orderVec[!(orderVec %in% lg2Order)]

orderVec <- c(
    orderVec,
    sample.order,
    lg2Order
)

Obio@databaseTable <- Obio@databaseTable[,orderVec]

## Get relevant colors ##
dfColor <- unique(Obio@dfDesign[,c("sample.id", "sample.group_color")])
row.names(dfColor) <- dfColor$sample.id

sample.colors <- sample.order
sample.colors <- gsub("_TPM", "", sample.colors)
sample.colors <- gsub("norm_counts_", "", sample.colors)

dfColor <- dfColor[sample.colors, ]

sample.colors <- as.vector(dfColor$sample.group_color)

#Obio@dfDesign[["ts_color"]] <- "#FF0000"
#Obio@dfDesign[Obio@dfDesign$dataseries == "Neuron", "ts_color"] <- "#009900"


#color.groups <- hue_pal()(length(color.groups))

#dfDesign[["sample_color"]] <- "orange"

# library(RColorBrewer)
# selcol <- colorRampPalette(brewer.pal(12,"Set3"))


#for (i in 1:length(color.groups)){
#    sample.colors[grep(color.groups[i], sample.colors)] <- color.sel[i]
#}

#dfGroupColors <- unique(data.frame(sample.color.order, sample.colors))
#names(dfGroupColors) <- gsub("sample.colors", "sample.group_color",names(dfGroupColors))

#Obio@dfDesign <- merge(Obio@dfDesign, dfGroupColors, by.x = "sample.group", "sample.color.order")


if (!exists("gsea.cat.lines")){
    gsea.cat.lines <- ""
    downloadCatEnrichmentFNxlsx <- ""



} else {
    downloadCatEnrichmentFNxlsx <- paste0("outputs/", Obio@parameterList$project.code, ".enriched.categories.txt")

    gseaList <- list(
        "enriched_categories_table" = paste0(
            "p",
            substr(Obio@parameterList$project_id, 4,100),
            "_enriched_categories_table"
        ),
        "NES_array" = "array('NES_Skin_Card14KI_vs_Skin_WT');",
        "padj_array" = "array('padj_Skin_Card14KI_vs_Skin_WT');",
        "image_link_array" = "array('contrast_1_image_link');"
    )
}

if (!exists("timecourse.cat.lines")){
    timecourse.cat.lines <- NA
}

#library(SBwebtools)
setwd(Obio@parameterList$localWorkDir)

#library(SBwebtools)
webSiteDir <- Obio@parameterList$localWorkDir

project.code <- Obio@parameterList$project_id

setwd(Obio@parameterList$localWorkDir)
# create.website.parameters(
#     df.data = Obio@databaseTable,
#     gene.id.column = Obio@parameterList$geneIDcolumn,
#     ptm.colum = "",
#     lab_id = "ref",
#     user_ids = Obio@parameterList$user_ids,
#     project_id = Obio@parameterList$project_id,
#     download_result_table = paste0("outputs/", project.code, ".result.table.txt"),
#     download_cat_enrichment_table = downloadCatEnrichmentFNxlsx,
#     database = Obio@parameterList$primDataDB,
#     reference_categories_db = "reference_categories_db_new",
#     labname = Obio@parameterList$labname,
#     rnaseqdbTableName = Obio@parameterList$rnaseqdbTableName,
#     lab.categories.table = Obio@parameterList$lab.categories.table,
#     sample.order = sample.order, #set to "" to go with default sorting
#     count.sample.colors = sample.colors,
#     count.table.headline = Obio@parameterList$count.table.headline,
#     count.table.sidelabel = Obio@parameterList$count.table.sidelabel,
#     webSiteDir = webSiteDir,
#     heamap.headline.text = Obio@parameterList$heamap.headline.text,
#     upper_heatmap_limit = 3,
#     lower_heatmap_limit = -3,
#     slider.selection.name = "logFC",
#     presentation.file = paste0("outputs/", Obio@parameterList$project_id, ".project.presentation.pptx"),
#     number_of_slides = 7,
#     default.sequence = "",
#     use.logFC.columns.for.heatmap = FALSE,
#     peptide.view.link = "",
#     create.2d.scatterplot.button = TRUE,
#     low_highlight = -1,
#     high_highlight =1,
#     display.qc = TRUE,
#     display.pca = TRUE,
#     display.report = TRUE,
#     pca.table.name = Obio@parameterList$PCAdbTableName,
#     gsea.cat.lines = gsea.cat.lines,
#     timecourse.cat.lines = timecourse.cat.lines,
#     venn.slider.selector.strings = c(
#         "_logFC",
#         "contrast_1_padj"
#     ),
#     plot.selection.strings = c(
#         "_logFC",
#         "_PCA_",
#         "_lg10p",
#         "contrast_P_PCA_estimatePCA",
#         "lg2BaseMean"
#     ), #NA, #strings to grep from col names for plot display
#     plate.view.db.table = NA,
#     plate.view.column.vec = NA,
#     cat.seletion.table.vec = c(Obio@parameterList$cat.ref.db, Obio@parameterList$cat.ref.db.table)
# )
#
# ###############################################################################
# ## Assemble microwebsite                                                     ##
#
# mWVec <- c(
#     "cp -r ../GSEA/enrichment_plots/ .",
#     "cp -r ../RSEM/Ensembl/multiqc_report.html .",
#     "cp -r /camp/stp/babs/working/boeings/Stefan/protocol_files/github/boeings/biologic/GD_bulkRNA_seq/* .",
#     paste0("mv outputs/p999.project.presentation.pptx outputs/", project.code, ".project.presentation.pptx"),
#     "cp ../../outputs/*xl* ./outputs/",
#     "cp ../DESeq2/MA* ./outputs/",
#     "cp ../DESeq2/PCA* ./outputs/",
#     "cp ../heatmap.col.* ./outputs/",
#     "cp ../DESeq2/covariance.plot.* ./report_figures"
# )
#
# setwd(paste0(Obio@parameterList$localWorkDir, "/", Obio@parameterList$project_id))
# sink("assembleMicrowebsite.sh")
# for (i in 1:length(mWVec)){
#     cat(mWVec[i]);cat("\n");
# }
# sink()

## Done assembling microwebsite                                              ##
###############################################################################

###############################################################################
## Deploy next generation website                                            ##



###############################################################################
## Do default Venn selection                                                 ##
defaultVennSel <- names(Obio@databaseTable)
defaultVennSel <- c(
    defaultVennSel[grep("contrast_", defaultVennSel)],
    defaultVennSel[grep("add_venn_", defaultVennSel)],
    defaultVennSel[grep("_TsCor_", defaultVennSel)]

)

rmVec <- c(
    grep("contrast_P_", defaultVennSel),
    grep("_lg2BaseMean_", defaultVennSel),
    grep("_padj_LRT_", defaultVennSel),
    grep("_lg10p_", defaultVennSel),
    grep("_stat_", defaultVennSel),
    grep("pValueCor_", defaultVennSel)
)

if (length(rmVec) > 0){
    defaultVennSel <- defaultVennSel[-rmVec]
}

## Add lrt
defaultVennSel <- c(
    defaultVennSel,
    names(Obio@databaseTable)[grep("contrast_L_lg10p", names(Obio@databaseTable))]
)

## Done with LRT selection                                                   ##
###############################################################################

###############################################################################
## Create default plot selection                                             ##
defaults <- sort(names(Obio@databaseTable)[grep("contrast_", names(Obio@databaseTable))])

xdefault <- defaults[grep("contrast_1_logFC", defaults)]
if (length(xdefault) == 0){
    xdefault <- defaults[1]
}

ydefault <- defaults[grep("contrast_1_lg10p", defaults)]
if (length(ydefault) == 0){
    ydefault <- defaults[2]
}



plotSelection <- c(
    names(Obio@databaseTable)[grep("_logFC_",names(Obio@databaseTable))],
    names(Obio@databaseTable)[grep("contrast_P_PCA_estimatePCA",names(Obio@databaseTable))],

    names(Obio@databaseTable)[grep("_lg10p_",names(Obio@databaseTable))],

    names(Obio@databaseTable)[grep("lg2BaseMean",names(Obio@databaseTable))],
    names(Obio@databaseTable)[grep("corCoef",names(Obio@databaseTable))]
)

chunk1 <- plotSelection[grep("_logFC_", plotSelection)]
chunkh2 <- plotSelection[grep("_lg10p_", plotSelection)]
chunk2a <-chunkh2[grep("_lg10p_PCA", chunkh2)]
chunk2b <- chunkh2[!(chunkh2  %in% chunk2a)]

chnk2c <- plotSelection[grep("_corCoef", plotSelection)]

chunk3 <- plotSelection[grep("_PCA_est", plotSelection)]

chunk4 <- plotSelection[grep("_lg2BaseMean_", plotSelection)]
# rmVec <- grep("_LRT_", chunk4)
# if (length(rmVec) > 0){
#     chunk4 <- chunk4[-rmVec]
# }

plotSelectionNew <- c(
    chunk1,
    chunk2b,
    chnk2c,
    chunk3,
    chunk2a,
    chunk4
)

plotAdd <- plotSelection[!(plotSelection %in% plotSelectionNew)]

plotSelection <- c(
    plotSelectionNew,
    plotAdd
)


## Done creating default plot selection                                      ##
###############################################################################

pos <- grep("designTScol", names(Obio@parameterList))
if (length(pos) == 1){
    timepoint <- Obio@parameterList$designTScol
} else {
    timepoint <- NULL
}

createSettingsFile(
    obj = Obio,
    df.data = Obio@databaseTable,
    defaultXcolName = xdefault,
    defaultYcolName = ydefault,
    timepointName = timepoint,
    primDataTable = Obio@parameterList$rnaseqdbTableName,
    pcaDbTable = Obio@parameterList$PCAdbTableName,
    sample.order = sample.order, #set to "" to go with default sorting
    heatmapSampleOrder = "",
    sample.names = "", # default is sample.order
    count.sample.colors = sample.colors,
    ptm.colum = "",
    count.table.headline = Obio@parameterList$count.table.headline,
    count.table.sidelabel = Obio@parameterList$count.table.sidelabel,
    venn.slider.selector.strings = defaultVennSel,
    plot.selection.strings = plotSelection,
    webSiteDir = paste0(hpc.mount, "Stefan/protocol_files/github/biologic/src/experiments/"),
    upper_heatmap_limit = 3,
    lower_heatmap_limit = -3,
    heamap.headline.text = Obio@parameterList$heamap.headline.text,
    project_id = Obio@parameterList$project_id
)

setwd(Obio@parameterList$localWorkDir)
## Done deploying next generation website                                    ##
###############################################################################



###########################################
# Do multiQC and add to QC section
###########################################
# run
# >> module use /camp/stp/babs/working/software/modules/all
# >> module load multiqc/0.9-2016b-Python-2.7.12

## Copy mga.tsv file into the working directory ##
# print(
#     paste0(
#         "cp ",
#         dataFolder,
#         "/mga.tsv ",
#         workdir
#     )
# )
#
# # run in top project directory:
# # >> multiqc .
# # multiqc /camp/stp/babs/working/boeings/Projects/103_VTL_ES_RNA_seq_BAFF_timecourse_hs/workdir/RSEM/Ensembl /camp/stp/babs/working/boeings/Projects/103_VTL_ES_RNA_seq_BAFF_timecourse_hs/workdir/RSEM/Ensembl/RNAseQC /camp/stp/babs/working/boeings/Projects/103_VTL_ES_RNA_seq_BAFF_timecourse_hs/workdir/logs /camp/stp/babs/working/boeings/Projects/103_VTL_ES_RNA_seq_BAFF_timecourse_hs/workdir/FASTQC_stranded
# print("module purge; module use /camp/stp/babs/working/software/modules/all; module load multiqc/1.3-2016b-Python-2.7.12")
# print("multiqc --config /camp/stp/babs/working/escudem/bin/multiqc_config.yaml .")
#
# print(paste0("multiqc . -o ", webSiteDir))
# print(paste0("multiqc . -o ", workdir, "outputs"))
#
# ## Transfer multiqc_report.html to outputs ##
#
# ## list directories to include in multiqc search
# #paste0(workdir, "logs")
# paste0(workdir, "RSEM/Ensembl")
# paste0(workdir, "RSEM/Ensembl/RNAseQC")
# paste0(workdir, "FASTQC")


#[61] knitr_1.15.1


###############################################################################
## Create report for this analysis                                           ##

## Use templateReportRNASeq.html and customize to this analysi               ##
###############################################################################

# listExistingProjects(
#     user     = "boeings",
#     password = db.pwd,
#     host     = "www.biologic-db.org",
#     dbname = "reference_categories_db_new",
#     dbtable = "project_description_table"
# )
#
# ###############################################################################
# ## Create new project if necessary                                           ##
#
# createNewProject(
#     dbname = "reference_categories_db_new",
#     dbtable = "project_description_table",
#     password = db.pwd,
#     project_name = "Mrtf_Srf_regulated_cytoskeletal_dynamics",
#     project_lab = ";Treisman;",
#     project_description = "Characterization of Mrtf and Srf transcription programs in the regulation of cytoskeletal dynamics."
# )

## Done creating new project                                                 ##
###############################################################################

###############################################################################
## List project in projects table                                            ##

# addProject2ProjectTable(
#     dbname = "reference_categories_db_new",
#     dbtable = "project_db_table",
#     password = db.pwd,
#     experiment_id= project_id,
#     experiment_question = "What are the potential target genes of YAP and TAZ in palate development? - <b>Lab plus literature data</b>",
#     experiment_description = "<b>RNA Seq analysis</b>. Examination of Yap Taz cartilage knockout animals at one day prior to birth has demonstrated the presence of a number of cleft palate phenotypes, including a failure of palatal shelf elevation. In addition, a developmental timeline derrived from the literature RNA-Seq experiment GSE66679 is examined. ",
#     experiment_owner = ";Hannah Vanyai;",
#     experiment_lab = paste0(";",labname,";"),
#     experiment_viewers = paste0(";", paste(user_ids, collapse = ";"),";"),
#     experiment_project = ";YAP_TAZ_in_Palate_Development;",
#     experiment_type = experiment.type,
#     experiment_details = "Experimental details will follow here",
#     experiment_x_coordinate = ";siCtrl;siTarget;",
#     experiment_x_coordinate_unit = "condition",
#     experiment_link = paste0(
#         "<a href='../",
#         project_id,
#         "/report.php' class='btn btn-success btn-lg' role='button'>Primary Data &raquo;</a>"
#     ),
#     experiment_title = ""
# )
## Done listing project                                                      ##
###############################################################################

##########
## Done ##
##########

###############################################################################
## Assemble documentation rmd script                                         ##

##                                                                           ##
###############################################################################


## End of template                                                           ##
###############################################################################

###############################################################################
## Add environment to documentation                                          ##

environmentVec <- c(
    "################################################################################",
    "## R environment                                                              ##",
    capture.output(sessionInfo()),
    "## End R environment                                                          ##",
    "################################################################################"
)

Obio <- add2vec(
    obj = Obio,
    slot_name = "scriptVec",
    value = environmentVec
)


## Done adding environment                                                   ##
###############################################################################

###############################################################################
## Upload reference gene categories                                          ##




## Done uploading reference gene categories                                  ##
###############################################################################
```

```{r, echo=TRUE, eval=TRUE, warning=FALSE, results="asis"}
###############################################################################
## Save Obio Object                                                          ##


save(Obio,
     file = paste0(
         Obio@parameterList$localWorkDir,
         Obio@parameterList$project_id,
         ".bioLOGIC.Robj"
     )
)

print("R bioLOGIC single cell object initialized.")

##                                                                           ##
###############################################################################

###############################################################################
## Run alignment shell scripts                                               ##


```

